flink-table/flink-table-api-java-bridge/src/test/java/org/apache/flink/table/factories/DataGenTableSourceFactoryTest.java

```diff
@@ -297,10 +295,104 @@ void testVariableLengthDataGeneration() throws Exception {
+            assertThat(row.getString(2).toString()).hasSize(25);
+            assertThat(row.getString(4).toString()).hasSize(9999);
+        }
+
+        assertThatThrownBy(
+                        () -> {
+                            descriptor.putString(
+                                    DataGenConnectorOptionsUtil.FIELDS
+                                            + ".f3."
+                                            + DataGenConnectorOptionsUtil.KIND,
+                                    DataGenConnectorOptionsUtil.RANDOM);
+                            descriptor.putLong(
+                                    DataGenConnectorOptionsUtil.FIELDS
+                                            + ".f3."
+                                            + DataGenConnectorOptionsUtil.LENGTH,
+                                    21);
+
+                            runGenerator(LENGTH_CONSTRAINED_SCHEMA, descriptor);
+                        })
+                .satisfies(
+                        anyCauseMatches(
+                                ValidationException.class,
+                                "Custom length '21' for variable-length type (VARCHAR/STRING/VARBINARY/BYTES) field 'f3' should be shorter than '20' defined in the schema."));
+    }
+
+    @Test
+    void testFixedLengthDataType() throws Exception {
+        DescriptorProperties descriptor = new DescriptorProperties();
+        final int rowsNumber = 200;
+        descriptor.putString(FactoryUtil.CONNECTOR.key(), "datagen");
+        descriptor.putLong(DataGenConnectorOptions.NUMBER_OF_ROWS.key(), rowsNumber);
+
+        List<RowData> results = runGenerator(LENGTH_CONSTRAINED_SCHEMA, descriptor);
+        assertThat(results).hasSize(rowsNumber);
+
+        for (RowData row : results) {
+            assertThat(row.getString(0).toString()).hasSize(50);
+            assertThat(row.getBinary(1)).hasSize(40);
+        }
+
+        assertThatThrownBy(
+                        () -> {
+                            descriptor.putString(
+                                    DataGenConnectorOptionsUtil.FIELDS
+                                            + ".f0."
+                                            + DataGenConnectorOptionsUtil.KIND,
+                                    DataGenConnectorOptionsUtil.RANDOM);
+                            descriptor.putLong(
+                                    DataGenConnectorOptionsUtil.FIELDS
+                                            + ".f0."
+                                            + DataGenConnectorOptionsUtil.LENGTH,
+                                    20);
+
+                            runGenerator(LENGTH_CONSTRAINED_SCHEMA, descriptor);
+                        })
+                .satisfies(
+                        anyCauseMatches(
+                                ValidationException.class,
+                                "Custom length for fixed-length type (CHAR/BINARY) field 'f0' is not supported."));
+    }
+
```

Comment:
Nit: I feel like the test is a little bit verbose; maybe we can simplify it a little bit. 
E.g.

extract a util
```java
private void assertException(
            ResolvedSchema schema,
            DescriptorProperties descriptor,
            String fieldName,
            @Nullable Integer len,
            @Nullable Boolean varLen,
            String expectedMessage) {
        assertThatThrownBy(
                        () -> {
                            descriptor.putString(
                                    String.join(
                                            ".",
                                            DataGenConnectorOptionsUtil.FIELDS,
                                            fieldName,
                                            DataGenConnectorOptionsUtil.KIND),
                                    DataGenConnectorOptionsUtil.RANDOM);
                            if (len != null) {
                                descriptor.putLong(
                                        String.join(
                                                ".",
                                                DataGenConnectorOptionsUtil.FIELDS,
                                                fieldName,
                                                DataGenConnectorOptionsUtil.LENGTH),
                                        len);
                            }
                            if (varLen != null) {
                                descriptor.putBoolean(
                                        String.join(
                                                ".",
                                                DataGenConnectorOptionsUtil.FIELDS,
                                                fieldName,
                                                DataGenConnectorOptionsUtil.VAR_LEN),
                                        varLen);
                            }

                            runGenerator(schema, descriptor);
                        })
                .satisfies(anyCauseMatches(ValidationException.class, expectedMessage));
```

then
```java
assertException(
                LENGTH_CONSTRAINED_SCHEMA,
                descriptor,
                "f0",
                20,
                null,
                "Custom length for fixed-length type (CHAR/BINARY) field 'f0' is not supported.");


assertException(
                LENGTH_CONSTRAINED_SCHEMA,
                descriptor,
                "f3",
                21,
                null,
                "Custom length '21' for variable-length type (VARCHAR/STRING/VARBINARY/BYTES) field 'f3' should be shorter than '20' defined in the schema.");


assertException(
                schema,
                descriptor,
                "f4",
                null,
                true,
                String.format(
                        "Only supports specifying '%s' option for variable-length types (VARCHAR/STRING/VARBINARY/BYTES). The type of field '%s' is not within this range.",
                        DataGenConnectorOptions.FIELD_VAR_LEN.key(), "f4"));
```

Reply:
good idea! how about using `assertLenRelatedException` instead of `assertException` ?